docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,321 @@
+      - [Implement within the cloud providers instead of Cluster API](#implement-within-the-cloud-providers-instead-of-cluster-api)
+      - [Implement as authentication webhook, as per aws-iam-authenticator (Amazon EKS)](#implement-as-authentication-webhook-as-per-aws-iam-authenticator-amazon-eks)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+    - [Test Plan [optional]](#test-plan-optional)
+    - [Graduation Criteria [optional]](#graduation-criteria-optional)
+      - [Graduation to beta](#graduation-to-beta)
+      - [Graduation to GA](#graduation-to-ga)
+    - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+
+- **OID:** Object Identifier defined by the International Telecommunications Union and used in PKI to identify attributes on certificates.
+
+- **PKI:** Public Key Infrastructure
+
+- **TPM:** Trusted Platform Module (TPM) is a specification defined by the Trusted Computing Group (TCG) that allows hosts to attest to their identity via PKI and a secure crypto-processor which may either be a separate chip, built into the CPU or a virtual device provided by the hypervisor.
+
+
+## Summary
+
+This proposal outlines a method to secure node registration within Cluster API, to solve 2 primary problems:
+- Solve a class of attacks involving node impersonation allowing an attacker to access secrets and volumes they shouldn’t by using hardware attestation of node identity.
+- Reduce kubeadm token reuse in MachinePools where the cloud provider does not support continuous update of the bootstrap userdata without creating new cloud provider specific machinepool resources (e.g. AWS Launch Configurations).
+
+This node attestation mechanism will be optional in the initial implementation, and can potentially be used independently of Cluster API.
+
+## Motivation
+
+Cluster API default core components are largely reliant on kubeadm for cluster bootstrapping and node registration. Kubeadm is a platform-agnostic command line tool designed to assist users to bootstrap Kubernetes clusters, and is used as a building block in Cluster API.
+Because kubeadm is platform-independent and is intended to provide an “easy path” to cluster bootstrapping,  there are a number of inherent design decisions that limit the overall security of the provisioned cluster:
+- Kubeadm uses TLS bootstrapping for node registration, however the default workflow used by Cluster API uses bootstrap token which allow registration as arbitrary node names.
+    - When used in this mode, Kubeadm essentially does “client-side validation” to prevent node hijacking, but this does not mean the token cannot be reused by an attacker within the lifetime of the token to perform a hijack.
+    - Cluster API cannot scope a token down to a specific node, because bootstrap providers, nor most infrastructure providers know the identity of the node ahead of time.
```

Comment:
```suggestion
    - Cluster API cannot scope a token down to a specific node, because neither bootstrap providers, nor most infrastructure providers know the identity of the node ahead of time.
```